drivers/interrupt_controller/intc_miwu.c

```diff
@@ -185,63 +185,61 @@ int npcx_miwu_interrupt_configure(const struct npcx_wui *wui,
 {
 	const uint32_t base = DRV_CONFIG(miwu_devs[wui->table])->base;
 	uint8_t pmask = BIT(wui->bit);
-
-	if (mode == NPCX_MIWU_MODE_DISABLED) {
-		/* Clear MIWU enable bit */
-		NPCX_WKEN(base, wui->group) &= ~pmask;
+	int ret = 0;
+
+	/* Handle interrupt for level trigger */
+	if (mode == NPCX_MIWU_MODE_LEVEL) {
+		/* Set detection mode to level */
+		NPCX_WKMOD(base, wui->group) |= pmask;
+		switch (trig) {
+		/* Enable interrupting on level high */
+		case NPCX_MIWU_TRIG_HIGH:
+			NPCX_WKEDG(base, wui->group) &= ~pmask;
+			break;
+		/* Enable interrupting on level low */
+		case NPCX_MIWU_TRIG_LOW:
+			NPCX_WKEDG(base, wui->group) |= pmask;
+			break;
+		default:
+			ret = -EINVAL;
+			break;
+		}
+	/* Handle interrupt for edge trigger */
 	} else {
-		/* Handle interrupt for level trigger */
-		if (mode == NPCX_MIWU_MODE_LEVEL) {
-			/* Set detection mode to level */
-			NPCX_WKMOD(base, wui->group) |= pmask;
-			switch (trig) {
-			/* Enable interrupting on level high */
-			case NPCX_MIWU_TRIG_HIGH:
-				NPCX_WKEDG(base, wui->group) &= ~pmask;
-				break;
-			/* Enable interrupting on level low */
-			case NPCX_MIWU_TRIG_LOW:
-				NPCX_WKEDG(base, wui->group) |= pmask;
-				break;
-			default:
-				return -EINVAL;
-			}
-		/* Handle interrupt for edge trigger */
-		} else {
-			/* Set detection mode to edge */
-			NPCX_WKMOD(base, wui->group) &= ~pmask;
-			switch (trig) {
-			/* Handle interrupting on falling edge */
-			case NPCX_MIWU_TRIG_LOW:
-				NPCX_WKAEDG(base, wui->group) &= ~pmask;
-				NPCX_WKEDG(base, wui->group) |= pmask;
-				break;
-			/* Handle interrupting on rising edge */
-			case NPCX_MIWU_TRIG_HIGH:
-				NPCX_WKAEDG(base, wui->group) &= ~pmask;
-				NPCX_WKEDG(base, wui->group) &= ~pmask;
-				break;
-			/* Handle interrupting on both edges */
-			case NPCX_MIWU_TRIG_BOTH:
-				/* Enable any edge */
-				NPCX_WKAEDG(base, wui->group) |= pmask;
-				break;
-			default:
-				return -EINVAL;
-			}
+		/* Set detection mode to edge */
+		NPCX_WKMOD(base, wui->group) &= ~pmask;
+		switch (trig) {
+		/* Handle interrupting on falling edge */
+		case NPCX_MIWU_TRIG_LOW:
+			NPCX_WKAEDG(base, wui->group) &= ~pmask;
+			NPCX_WKEDG(base, wui->group) |= pmask;
+			break;
+		/* Handle interrupting on rising edge */
+		case NPCX_MIWU_TRIG_HIGH:
+			NPCX_WKAEDG(base, wui->group) &= ~pmask;
+			NPCX_WKEDG(base, wui->group) &= ~pmask;
+			break;
+		/* Handle interrupting on both edges */
+		case NPCX_MIWU_TRIG_BOTH:
+			/* Enable any edge */
+			NPCX_WKAEDG(base, wui->group) |= pmask;
+			break;
+		default:
+			ret = -EINVAL;
+			break;
 		}
+	}
 
-		/* Enable wake-up input sources */
-		NPCX_WKINEN(base, wui->group) |= pmask;
+	/* Enable wake-up input sources */
+	NPCX_WKINEN(base, wui->group) |= pmask;
```

Comment:
Should these changes be conditional on a non-negative `ret`?

AFAICT `NPCX_MIWU_MODE_DISABLED` is no longer part of the logic here; it used to clear the pmask.  Even if this is not an issue because the GPIO call sites already disabled it, I suspect it should be handled here.

Reply:
I'm sorry about our ambiguous register name. (I try to add more explanations in each PR possibly without npcx datasheet.)
NPCX_WKEN is "Wake-Up Enable  Register" which is used for enabling/disabling to generate a wake-up signal or interrupt request.
NPCX_WKINEN is "Wake-Up Input Enable x Register" which is used for connecting io-pad and MIWU modules. If it is false, the pin can be left unconnected (i.e., floating).
You need to turn both WKEN and WKINEN on for handling a gpio event. Disabling one of them means turns the interrupt off. But I look this change again. The driver should return -EINVAL directly if the configuration setting is wrong. Then the gpio driver will show an error log and return error code directly.

Reply:
OK, I think I understand what this is doing better.

I suspect the commit should be split into two pieces:

* The change to gpio_npcx, which now uses a simpler configuration approach that turns interrupts off before configuring them;
* The change to intc_npcx, which (AFAICT) is no longer responsible for configuring `NPCX_WK` to disable interrupts, because it expects that to have been done before it's called.

I think the current commit message describes the gpio changes, but nothing describes the intc ones.

If I've reconstructed that correctly, please update this.  Otherwise please update it to explain why both files need to be touched in the same commit (e.g. they're tightly coupled and the system won't work without both).

Also (if I've described the change correctly) have you confirmed that all other callers of `npcx_miwu_interrupt_configure()` have disabled interrupts before it's called?  Or maybe the disable should be pushed down into that module?

Reply:
> OK, I think I understand what this is doing better.
> 
> I suspect the commit should be split into two pieces:
> 
> * The change to gpio_npcx, which now uses a simpler configuration approach that turns interrupts off before configuring them;
> * The change to intc_npcx, which (AFAICT) is no longer responsible for configuring `NPCX_WK` to disable interrupts, because it expects that to have been done before it's called.
> 
> I think the current commit message describes the gpio changes, but nothing describes the intc ones.
> 
> If I've reconstructed that correctly, please update this. Otherwise please update it to explain why both files need to be touched in the same commit (e.g. they're tightly coupled and the system won't work without both).
>

I will split it into two CLs for better explanations. Thanks.
  
> Also (if I've described the change correctly) have you confirmed that all other callers of `npcx_miwu_interrupt_configure()` have disabled interrupts before it's called? Or maybe the disable should be pushed down into that module?

The default value of WK_EN is zero. The other functions call npcx_miwu_interrupt_configure() during driver init function. Only the gpio driver provides an interface to configure interrupt manually. That's why we saw the side-effect when users configure the same pin twice. But it is safer to add a check for WK_EN in miwu driver.

Reply:
Done.